### PR TITLE
Fix snack bar behavior

### PR DESCRIPTION
**Pull request by janziemba**
35264955-e4b1e558-001e-11e8-89f2-27ebbe5baeb5

### CODE REVIEW

**Comment by xotahal** [comment]
Why did you remove `visibility` prop?

**Comment by janziemba** [comment]
Because `nextProps.visible` and `this.props.visible` were always the same so the `show` and `setHideTimer` functions in `componentWillReceiveProps` has never been called. Also, why would you change visibility from the outside? This component hides after given timeout. Or could you please give an example how this should work?

**Comment by xotahal** [comment]
@janziemba well, I think it's all wrong :D I think this component can't have sth like bottomNavigation at all ... we should be able to set some bottom margin, but not bottomNavigation .. it's bad name for this i think ...

The second think is visible prop. There is a reason to have it. Because you should be able to chose if you want snack bar which is hide automatically, or by visible props ..

You can see RN's Modal component that it's pretty normal to have visible prop - https://facebook.github.io/react-native/docs/modal.html#visible

I wouldn't say it is useless ..

**Comment by HofmannZ** [comment]
@xotahal When I wrote this the idea was that the snackbar would alway be mounted. I think @janziemba is using a implementation where you only mount the snack bar when you need it.

The reason why I chose to have the snackbar always mounted is the following:
- It would allow the snackbar to animate smoothly, I saw some frames drop when mounting and directly animation the component.
- It would make it possible to add an event queue, where multiple messages are displayed in a specific sequence, see the [MD Guidelines](https://material.io/guidelines/components/snackbars-toasts.html#snackbars-toasts-usage) (Consecutive snackbars).

I do think that it should use native animations on iOS as well, witch was not supported at the time.

**Comment by xotahal** [comment]
@HofmannZ Yep native animations are ok ... make sense .. we should have it always mounted .. 

Then it can be hide automatically by timer, or by set the visible prop ...

Let's say there will be a `hideAfter: milliseconds` prop .. Then we now we should hide the snackbar automatically after `milliseconds` ... if it's not presented it will be open till the visible is set to false ...

By default visible will be set to false ... To give the user option to animate snackbar after it is mounted ..

In this set up we can use like "always mounted" or "mount it when we need it" ...

Hmmm? What do you think guys?

**Comment by janziemba** [comment]
I have updated this PR.
What do you think about this guys?

untitled3

**Comment by HofmannZ** [comment]
@xotahal Well I think Snackbars should auto hide by default be cause that's clearly written in the guidelines. Toasts should stay on the screen, so maybe make a wrapper component called toast and uses a Snackbar component under the hood.

I think a `hideAfter: milliseconds` hide after prop would a great way to solve it. I just peaked at the [official material components repository](https://github.com/material-components/material-components-web/blob/master/packages/mdc-snackbar), and they call it the `timeout`  property with a default of 2750.

@janziemba As per guidelines a Snackbar cannot cover the FAB. So I think we need some way to communicate between those components.

What do you guys think?

File: src/Snackbar/Snackbar.react.js [hunk 1]
```diff
@@ -86,24 +90,27 @@ class Snackbar extends PureComponent {
         const styles = getStyles(props, context);
         this.state = {
             styles,
-            moveAnimated: new Animated.Value(0),
+            moveAnimated: new Animated.Value(48),
```
Comment (xotahal) [inline src/Snackbar/Snackbar.react.js]: What's the 48? It should have been height from styles, right?

It should be set to 0 if visible === false, or maybe -height?

File: src/Snackbar/Snackbar.react.js [hunk 2]
```diff
@@ -123,40 +130,32 @@ class Snackbar extends PureComponent {
...
             duration: 225,
             easing: Easing.bezier(0.0, 0.0, 0.2, 1),
             useNativeDriver: true,
         }).start();
     }
 
     hide = () => {
-        const { moveAnimated, styles } = this.state;
-        Animated.timing(moveAnimated, {
-            toValue: (StyleSheet.flatten(styles.container).height),
```
Comment (xotahal) [inline src/Snackbar/Snackbar.react.js]: We should use this ..

**Comment by xotahal** [comment]
Ok, so let's set `timeout` prop to 2750 by default. And add possibility to set `timeout` prop to `null` which will cause that snackbar will be shown till the `visible` prop is set to false. Make sense?

We still can have a `visible` prop set to false by default. To be able to use the component in "always mounted" mode

**Comment by HofmannZ** [comment]
Yeah I agree, that way it’s best of both worlds 👍🏼
		
		

		Zino Hofmann - Founder | CTIO+31 6 29435375 
Zino App B.V.Levantkade 871019 MJ AmsterdamThe Netherlands 
DisclaimerThis email and any files transmitted with it are confidential and intended solely for the use of the individual or entity to whom they are addressed. If you have received this email in error please notify the system manager. Please note that any views or opinions presented in this email are solely those of the author and do not necessarily represent those of the company. Finally, the recipient should check this email and any attachments for the presence of viruses. The company accepts no liability for any damage caused by any virus transmitted by this email.
Sent from my iPhone
	




On Mon, Jan 29, 2018 at 10:19 PM +0100, "Jiří Otáhal" <notifications@github.com> wrote:












Ok, so let's set timeout prop to 2750 by default. And add possibility to set timeout prop to null which will cause that snackbar will be shown till the visible prop is set to false. Make sense?


We still can have a visible prop set to false by default. To be able to use the component in "always mounted" mode

**Comment by janziemba** [comment]
There has always been `timeout` prop with a default value of 2750.

So the only change is that `bottomNavigation` is now `bottomMargin` and we use native animations.

Is that correct?

**Comment by HofmannZ** [comment]
We did kinda drop support for a standard use case here. With the new implementation we need to pass explicitly pass the bottom navigation height rather that boolean. Although it's not the end of the world, we could could fix it with a prop like `interactWithBottomNavigation` witch would internally overwrite the value of `bottomMargin` with the bottomNavigation height from the theme.

Also this PR introduces some breaking changes, @xotahal maybe motion that in the release notes?

**Comment by janziemba** [comment]
Forget it, guys. I think I have used this component a wrong way. The original source code does what I need so there is no need to update it. Sorry.
Or would you like to continue working on this PR?

**Comment by HofmannZ** [comment]
@janziemba I can implement the changes tomorrow, I think it might actually be useful for some people. 

@xotahal What do you think?

**Comment by xotahal** [comment]
Yep, let's finish this when we've started already .. I will write a breaking changes ...

**Pull request closed without merging**